File: static/py/ui/timeline_container.py
"""
Timeline Container - Spatial layout manager for lesson components
Handles the visual timeline structure but NOT component logic/lifecycle
"""

from pyscript import document # type: ignore

class TimelineContainer:
    """Manages the spatial layout of lesson components in a timeline structure"""
    
    def __init__(self, container_id="lesson-container"):
        self.container_id = container_id
        self.timeline_id = f"{container_id}-timeline"
        self.error_slot_id = f"{container_id}-error-slot"
        self._setup_timeline_structure()
    
    def _setup_timeline_structure(self):
        """Create the basic timeline DOM structure"""
        container = document.getElementById(self.container_id)
        if not container:
            raise ValueError(f"Container {self.container_id} not found")
        
        # Clear any existing content
        container.innerHTML = ""
        
        # Create timeline wrapper
        timeline_html = f'''
        <div class="timeline-wrapper max-w-4xl mx-auto">
            <!-- Error slot (hidden by default) -->
            <div id="{self.error_slot_id}" class="hidden mb-6"></div>
            
            <!-- Main timeline -->
            <div id="{self.timeline_id}" class="timeline-track relative">
                <!-- Timeline line -->
                <div class="timeline-line absolute left-8 top-0 bottom-0 w-0.5 bg-gray-300"></div>
                
                <!-- Component slots will be added here -->
            </div>
        </div>
        '''
        container.innerHTML = timeline_html
    
    def add_component_slot(self, component_id, position="bottom"):
        """Add a spatial slot for a component - matches learn.html card styling"""
        timeline = document.getElementById(self.timeline_id)
        if not timeline:
            return False
        
        slot_html = f'''
        <div id="slot-{component_id}" class="timeline-item relative flex items-start">
            <!-- Timeline dot -->
            <div class="timeline-dot relative z-10 flex items-center justify-center w-4 h-4 bg-white dark:bg-gray-800 border-2 border-blue-500 rounded-full mt-6">
                <div class="w-1.5 h-1.5 bg-blue-500 rounded-full"></div>
            </div>
            
            <!-- Component content area - matches learn.html card styling -->
            <div id="component-{component_id}" class="component-content ml-6 flex-1 bg-white dark:bg-gray-800 rounded-xl shadow-lg border border-gray-200 dark:border-gray-700 p-6">
                <!-- Core will inject component here -->
                <div class="loading-placeholder text-center py-4">
                    <div class="flex items-center justify-center space-x-2 text-blue-600 dark:text-blue-400">
                        <svg class="animate-spin h-5 w-5" fill="none" viewBox="0 0 24 24">
                            <circle class="opacity-25" cx="12" cy="12" r="10" stroke="currentColor" stroke-width="4"></circle>
                            <path class="opacity-75" fill="currentColor" d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"></path>
                        </svg>
                        <span class="text-sm font-medium">Loading component...</span>
                    </div>
                </div>
            </div>
        </div>
        '''
        
        if position == "bottom":
            timeline.insertAdjacentHTML('beforeend', slot_html)
        else:
            timeline.insertAdjacentHTML('afterbegin', slot_html)
        
        return True
    
    def get_component_area(self, component_id):
        """Get the DOM element where core should inject the component"""
        return document.getElementById(f"component-{component_id}")
    
    def remove_component_slot(self, component_id):
        """Remove a component slot from the timeline"""
        slot = document.getElementById(f"slot-{component_id}")
        if slot:
            slot.remove()
            return True
        return False
    
    def update_component_status(self, component_id, status="active"):
        """Update visual state of a component slot (completed, active, locked)"""
        slot = document.getElementById(f"slot-{component_id}")
        if not slot:
            return False
        
        dot = slot.querySelector('.timeline-dot div')
        if not dot:
            return False
        
        # Reset classes
        dot.className = "w-2 h-2 rounded-full"
        
        # Apply status styling
        if status == "completed":
            dot.className += " bg-green-500"
            slot.querySelector('.timeline-dot').className = slot.querySelector('.timeline-dot').className.replace('border-blue-500', 'border-green-500')
        elif status == "active":
            dot.className += " bg-blue-500"
            slot.querySelector('.timeline-dot').className = slot.querySelector('.timeline-dot').className.replace('border-green-500', 'border-blue-500')
        elif status == "locked":
            dot.className += " bg-gray-400"
            slot.querySelector('.timeline-dot').className = slot.querySelector('.timeline-dot').className.replace('border-blue-500 border-green-500', 'border-gray-400')
        
        return True
    
    def get_error_slot(self):
        """Get the error display area for the error handler"""
        return document.getElementById(self.error_slot_id)
    
    def clear_timeline(self):
        """Remove all component slots"""
        timeline = document.getElementById(self.timeline_id)
        if timeline:
            # Keep the timeline line, remove component slots
            slots = timeline.querySelectorAll('.timeline-item')
            for slot in slots:
                slot.remove()
    
    def set_timeline_progress(self, completed_count, total_count):
        """Update the timeline line to show overall progress"""
        if total_count == 0:
            return
        
        progress_percent = (completed_count / total_count) * 100
        timeline = document.getElementById(self.timeline_id)
        if not timeline:
            return
        
        # Update timeline line with progress indicator
        timeline_line = timeline.querySelector('.timeline-line')
        if timeline_line:
            timeline_line.innerHTML = f'''
            <div class="progress-fill absolute top-0 left-0 w-full bg-green-400" 
                 style="height: {progress_percent}%; transition: height 0.3s ease;"></div>
            '''

File: src/ts/ui/timelineContainer.ts
// timelineContainer.ts - Simplified vertical timeline for component flow
// Just manages spatial layout - components render in order with consistent spacing

/**
 * Manages the spatial layout of lesson components in a simple vertical flow.
 * No decorative timeline elements - just provides slots for components to render into.
 *
 * Responsibilities:
 * - Create component slots in order
 * - Provide render areas for component interfaces
 * - Clear all slots on page navigation
 *
 * Does NOT handle:
 * - Component lifecycle (that's componentCoordinator)
 * - Visual state/status (components handle their own appearance)
 * - Progress tracking (that's in the data layer)
 */
export class TimelineContainer {
  private containerId: string;
  private timelineId: string;

  constructor(containerId: string = "lesson-container") {
    this.containerId = containerId;
    this.timelineId = `${containerId}-timeline`;
    this.setupTimelineStructure();
  }

  /**
   * Create the basic timeline DOM structure - just a centered container
   */
  private setupTimelineStructure(): void {
    const container = document.getElementById(this.containerId);
    if (!container) {
      throw new Error(`Container ${this.containerId} not found`);
    }

    // Simple centered container for vertical component flow
    container.innerHTML = `
        <div class="max-w-4xl lg:max-w-6xl mx-auto py-8 px-4">
            <div id="${this.timelineId}" class="space-y-8">
              <!-- Component slots will be added here -->
             </div>
        </div>
    `;

    console.log("✅ Timeline container initialized");
  }

  /**
   * Add a slot for a component to render into.
   * Creates a simple wrapper div with consistent spacing.
   *
   * @param componentId - Unique identifier for the component
   */
  addComponentSlot(componentId: number): void {
    const timeline = document.getElementById(this.timelineId);
    if (!timeline) {
      console.error("❌ Timeline not found, cannot add slot");
      return;
    }

    const slotHtml = `
            <div id="slot-${componentId}" class="component-slot">
                <div id="component-area-${componentId}" class="component-content">
                    <!-- Component interface renders here -->
                </div>
            </div>
        `;

    timeline.insertAdjacentHTML("beforeend", slotHtml);
    console.log(`📍 Added slot for component ${componentId}`);
  }

  /**
   * Get the DOM element where a component should render its content.
   * Component interfaces use this to know where to attach their UI.
   *
   * @param componentId - Component identifier
   * @returns The render area element, or null if not found
   */
  getComponentArea(componentId: number): HTMLElement | null {
    const area = document.getElementById(`component-area-${componentId}`);
    if (!area) {
      console.warn(`⚠️ Component area ${componentId} not found`);
    }
    return area;
  }

  /**
   * Remove all component slots from the timeline.
   * Called when navigating to a new page - clears everything for fresh start.
   */
  clearTimeline(): void {
    const timeline = document.getElementById(this.timelineId);
    if (timeline) {
      timeline.innerHTML = "";
      console.log("🧹 Timeline cleared of all component slots");
    }
  }

  /**
   * Get basic timeline statistics for debugging.
   *
   * @returns Object with timeline metrics
   */
  getTimelineStats(): {
    totalSlots: number;
    containerId: string;
    timelineId: string;
  } {
    const timeline = document.getElementById(this.timelineId);
    const totalSlots = timeline
      ? timeline.querySelectorAll(".component-slot").length
      : 0;

    return {
      totalSlots,
      containerId: this.containerId,
      timelineId: this.timelineId,
    };
  }
}

/**
 * Singleton instance - created once and shared across application.
 * Bootstrap initializes this before starting core.
 */
let timelineInstance: TimelineContainer | null = null;

/**
 * Initialize the singleton timeline instance.
 * Called once by bootstrap during initialization.
 *
 * @param containerId - ID of the container element (default: 'lesson-container')
 * @returns The timeline instance
 */
export function initializeTimeline(
  containerId: string = "lesson-container",
): TimelineContainer {
  if (timelineInstance) {
    console.warn(
      "⚠️ Timeline already initialized, returning existing instance",
    );
    return timelineInstance;
  }

  timelineInstance = new TimelineContainer(containerId);
  console.log("✅ Timeline singleton initialized");
  return timelineInstance;
}

/**
 * Get the singleton timeline instance.
 *
 * @throws Error if timeline not yet initialized
 * @returns The timeline instance
 */
export function getTimelineInstance(): TimelineContainer {
  if (!timelineInstance) {
    throw new Error(
      "Timeline not initialized. Call initializeTimeline() in bootstrap first.",
    );
  }
  return timelineInstance;
}
